destination/destination.go

```diff
@@ -0,0 +1,271 @@
+	"github.com/conduitio/conduit-commons/lang"
+	"github.com/conduitio/conduit-commons/opencdc"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+type Destination struct {
+	sdk.UnimplementedDestination
+
+	config     config.Config
+	sshClient  *ssh.Client
+	sftpClient *sftp.Client
+}
+
+func NewDestination() sdk.Destination {
+	return sdk.DestinationWithMiddleware(&Destination{}, sdk.DefaultDestinationMiddleware(
+		sdk.DestinationWithSchemaExtractionConfig{
+			PayloadEnabled: lang.Ptr(false),
+			KeyEnabled:     lang.Ptr(false),
+		},
+	)...)
+}
+
+func (d *Destination) Parameters() commonsConfig.Parameters {
+	return d.config.Parameters()
+}
+
+func (d *Destination) Configure(ctx context.Context, cfg commonsConfig.Config) error {
+	sdk.Logger(ctx).Info().Msg("Configuring Destination...")
+	err := sdk.Util.ParseConfig(ctx, cfg, &d.config, NewDestination().Parameters())
+	if err != nil {
+		return fmt.Errorf("invalid config: %w", err)
+	}
+
+	err = d.config.Validate()
+	if err != nil {
+		return fmt.Errorf("error validating configuration: %w", err)
+	}
+
+	return nil
+}
+
+func (d *Destination) Open(ctx context.Context) error {
+	sdk.Logger(ctx).Info().Msg("Opening a SFTP Destination...")
+	sshConfig, err := d.sshConfigAuth()
+	if err != nil {
+		return fmt.Errorf("failed to create SSH config: %w", err)
+	}
+
+	d.sshClient, err = ssh.Dial("tcp", d.config.Address, sshConfig)
+	if err != nil {
+		return fmt.Errorf("failed to dial SSH: %w", err)
+	}
+
+	d.sftpClient, err = sftp.NewClient(d.sshClient)
+	if err != nil {
+		d.sshClient.Close()
+		return fmt.Errorf("failed to create SFTP client: %w", err)
+	}
+
+	_, err = d.sftpClient.Stat(d.config.DirectoryPath)
+	if err != nil {
+		return fmt.Errorf("remote path does not exist: %w", err)
+	}
+
+	return nil
+}
+
+func (d *Destination) Write(_ context.Context, records []opencdc.Record) (int, error) {
+	for i, record := range records {
+		chunked, ok := record.Metadata["is_chunked"]
+		if ok && chunked == "true" {
+			err := d.handleChunkedRecord(record)
+			if err != nil {
+				return i, err
+			}
+			continue
+		}
+
+		filename, ok := record.Metadata["filename"]
+		if !ok {
+			filename = string(record.Key.Bytes())
+		}
+
+		err := d.uploadFile(filename, record.Payload.After.Bytes())
+		if err != nil {
+			return i, err
+		}
+	}
+
+	return len(records), nil
+}
+
+func (d *Destination) handleChunkedRecord(record opencdc.Record) error {
+	index, ok := record.Metadata["chunk_index"]
+	if !ok {
+		return NewInvalidChunkError("chunk_index not found")
+	}
+
+	totalChunks, ok := record.Metadata["total_chunks"]
+	if !ok {
+		return NewInvalidChunkError("total_chunk not found")
+	}
+
+	hash, ok := record.Metadata["hash"]
+	if !ok {
+		return NewInvalidChunkError("hash not found")
+	}
+
+	var remoteFile *sftp.File
+	var err error
+	path := fmt.Sprintf("%s/%s.tmp", d.config.DirectoryPath, hash)
```

Comment:
Currently, the SFTP source hashes the modification time, which means that if multiple files are being written, and at least 2 of those have the same modification time, the contents might get mixed up.

Reply:
Yes, the hash is now updated in the source to generate a unique hash with filename and modtime.